### ISSUE TITLE
Preserve colors in output of exec()

### ISSUE THREAD

**Issue by arturadib** [shelljs/shelljs#86]
See http://stackoverflow.com/a/14231570/201852

**Comment by EragonJ** [comment]
:+1:

**Comment by nilssolanki** [comment]
:+1: Please.

**Comment by philipplaeubli** [comment]
+1

**Comment by Nirazul** [comment]
+ 1

**Comment by sbmaxx** [comment]
+1

**Comment by JulianLaval** [comment]
+1

**Comment by vbud** [comment]
:+1:

**Comment by allenmanning** [comment]
:+1:

**Comment by vjpr** [comment]
+1

**Comment by yoyurec** [comment]
+1

**Comment by bmson** [comment]
:+1:

**Comment by nathanbrauer** [comment]
I got this working by telling the program to always send colors (in my case `composer --ansi`), but it would be great if it was native with no additional step needed.

**Comment by Loilo** [comment]
+1

**Comment by kazazor** [comment]
:+1:

**Comment by ariporad** [comment]
If someone want's to submit a PR, that would be fantastic! Otherwise, we'll try and get this in by `v1.0.0`.

**Comment by sedwards2009** [comment]
As already mentioned, some programs can be told to output colour using a flag. Many programs detect if they are talking to a TTY to decide whether they output colour and other terminal features, or just plain text. To make all programs think that they have a TTY/terminal connected you need to run the programs with the help of something like https://www.npmjs.com/package/pty.js instead of using exec() directly. This runs the program with a pseudo-TTY connected and the program just thinks it is a on a TTY.

**Comment by nfischer** [comment]
Thanks, @sedwards2009! We would unfortunately need to find some sort of Windows equivalent before we could use that, but that's a good start :+1:

**Comment by jurgob** [comment]
:+1:

**Comment by Permagate** [comment]
:+1:

**Comment by babinia** [comment]
:+1:

**Comment by TossShinHwa** [comment]
Any news for this issue?

**Comment by nfischer** [comment]
@TossShinHwa take a look at #524 and let me know if it preserves colors/spinners

**Comment by LucienLee** [comment]
The current workaround is passing parameter in `exec` to specify enable color manually. For example, `shell.exec('node yourscript.js --color always');` Basically, it's same as @nathanbrauer' method.

**Comment by danday74** [comment]
Another working example:

var command = 'git diff --color=always'

exec(command etc...)

**Comment by hstarorg** [comment]
@LucienLee @danday74 Thank you, It's useful in `gulp` too.
My code:
```js
let gulpfilePath = util.root('dev-package/build/gulpfile.js');
  let cwd = util.cwd();
  let gulpCmd = `gulp4 build --gulpfile ${gulpfilePath} --cwd ${cwd} --color=always`;
  exec(gulpCmd);
```

**Comment by alexreg** [comment]
This really calls out for a proper solution. The issue is 4 years old now. Any chance of having one, please? :-)

**Comment by dylanbhughes** [comment]
Bump

**Comment by johnunclesam** [comment]
+1

**Comment by SarathSantoshDamaraju** [comment]
@nfischer any update on this? I am working on a CLI project related to git and this is basically printing the git command results in a plain text even without git default colouring. And `--color=always` is not a perfect solution as it only applies for **diff** so not a perfect fit for my requirement.

**Comment by luisnaranjo733** [comment]
Also would love to see this happen. The --color=always approach only works sometimes, for programs that have that flag. Detox, for instance, doesn't support that option.

**Comment by luisnaranjo733** [comment]
Sweetness. Sounds like using the inherit option under the hood would solve this problem.

**Comment by iderets** [comment]
👍 It would be great to see it implemented

**Comment by a-x-** [comment]
OHHH, can we hope here? 
I rewrote some big script from bash to shelljs today and stuck with this bug 😫😭

**Comment by a-x-** [comment]
### Workaround summary
- there is something like `--color=always` on some utils
- a [pty.js](https://www.npmjs.com/package/pty.js) as referenced [here](https://github.com/shelljs/shelljs/issues/86#issuecomment-185221754)

**Comment by DanielRuf** [comment]
Bump.

**Comment by tenjojeremy** [comment]
Bump

**Comment by isuvorov** [comment]
+1

**Comment by distante** [comment]
Am I right to assume this is in discussion for development but not yet implemented?

**Comment by DanielRuf** [comment]
@distante as you can see this issue is still open so it might not be implemented. PRs are very welcome. This is opensource software so anyone can contribute.

**Comment by distante** [comment]
> @distante as you can see this issue is still open so it might not be implemented. PRs are very welcome. This is opensource software so anyone can contribute.

Yes, I am just do not understand this doc : https://docs.google.com/document/d/1UFm10TONaNWok3aEPzUP_OjZ6lEvwlYqyJBUcugLfso/edit#

**Comment by nfischer** [comment]
I believe color output is a potential outcome of that design. I haven't implemented that part of the design yet, but I'll update this issue when I do.

**Comment by jindrahm** [comment]
Adding option` {stdio: 'inherit'}` works for me, see https://stackoverflow.com/a/14231570.

**Comment by tangtangtangtangtang** [comment]
> I believe color output is a potential outcome of that design. I haven't implemented that part of the design yet, but I'll update this issue when I do.

still not implemented? what's the plan... is's 2022... @nfischer

**Comment by laterdayi** [comment]
@nfischer 

is there any progress

**Comment by jaikme** [comment]
Well, it's not very "clean", but you can try [FORCE_COLOR](https://nodejs.org/dist/latest-v18.x/docs/api/cli.html#force_color1-2-3) env.
```
shelljs.exec(cmd, {
  // ... other options
  stdio: [process.stdin, silent ? null : process.stdout, silent ? null : process.stderr],
  env: { ...process.env, FORCE_COLOR: true },
});
```
